src/gt4py/next/program_processors/runners/dace_fieldview/gtir_to_tasklet.py

```diff
@@ -167,16 +181,281 @@ def _visit_deref(self, node: gtir.FunCall) -> MemletExpr | ValueExpr:
+            offset_table_node,
+            sbs.Range.from_array(offset_table_node.desc(self.sdfg)),
+            tasklet_node,
+            "table",
+        )
+        if isinstance(offset_expr, MemletExpr):
+            self._add_entry_memlet_path(
+                offset_expr.node,
+                offset_expr.subset,
+                tasklet_node,
+                "offset",
+            )
+        else:
+            self.state.add_edge(
+                offset_expr.node,
+                None,
+                tasklet_node,
+                "offset",
+                dace.Memlet(data=offset_expr.node.data, subset="0"),
+            )
+
+        dtype = offset_table_node.desc(self.sdfg).dtype
+        return self._get_tasklet_result(dtype, tasklet_node, new_index_connector)
+
+    def _make_unstructured_shift(
+        self,
+        it: IteratorExpr,
+        connectivity: gtx_common.Connectivity,
+        offset_table_node: dace.nodes.AccessNode,
+        offset_expr: IteratorIndexExpr,
+    ) -> IteratorExpr:
+        """Implements shift in unstructured domain by means of a neighbor table."""
+        assert connectivity.neighbor_axis in it.dimensions
+        neighbor_dim = connectivity.neighbor_axis
+        assert neighbor_dim not in it.indices
+
+        origin_dim = connectivity.origin_axis
+        assert origin_dim in it.indices
+        origin_index = it.indices[origin_dim]
+        assert isinstance(origin_index, SymbolExpr)
+
+        if isinstance(offset_expr, SymbolExpr):
+            # use memlet to retrieve the neighbor index
+            shifted_indices = it.indices | {
+                neighbor_dim: MemletExpr(
+                    offset_table_node,
+                    sbs.Indices([origin_index.value, offset_expr.value]),
+                )
+            }
+        else:
+            # dynamic offset: we cannot use a memlet to retrieve the offset value, use a tasklet node
+            dynamic_offset_value = self._make_dynamic_neighbor_offset(
+                offset_expr, offset_table_node, origin_index
+            )
+
+            shifted_indices = it.indices | {neighbor_dim: dynamic_offset_value}
+
+        shifted_indices.pop(origin_dim)
+        return IteratorExpr(it.field, it.dimensions, shifted_indices)
+
+    def _visit_shift(self, node: gtir.FunCall) -> IteratorExpr:
+        shift_node = node.fun
+        assert isinstance(shift_node, gtir.FunCall)
+
+        # here we check the arguments to the `shift` builtin function: the offset provider and the offset value
+        head, tail = self._split_shift_args(shift_node.args)
+        if tail:
+            # we visit a multi-dimensional shift as recursive shift function calls, each returning a new iterator
+            it = self.visit(self._make_shift_for_rest(tail, node.args[0]))
+        else:
+            # the iterator to be shifted is the argument to the function node
+            it = self.visit(node.args[0])
+        assert isinstance(it, IteratorExpr)
```

Comment:
Maybe I'm completely wrong here, but it seems that this block is kind of canonicalizing the shift specification to a specific way to express it, so I would suggest to extract it as a function with a meaningful name.